src/EmpiricalModeDecomposition.jl

```diff
@@ -16,67 +17,62 @@ include("utils.jl")
 include("sift.jl")
 
 
-
-
 """
 Iterator for the Empirical Mode Decomposition.
-The time series values are an AbstractVector of type T,
-and the time positions are an AbstractVector of type U.
+The time series values are an AbstractVector of type `T`,
+and the time positions are an AbstractVector of type `U`.
 """
-struct EMDIterable{T<:AbstractVector,U<:AbstractVector}
+struct EMDIterable{T<:AbstractVector, U<:AbstractVector}
     yvec::T
     xvec::U
 end
 
 function iterate(iter::EMDIterable, imf_prev=(iter.yvec,false))
     if imf_prev[2]
-      return nothing
-    elseif sum(abs,imf_prev[1])>0.0 && !ismonotonic(imf_prev[1])
-      imf = sift(imf_prev[1],iter.xvec, 0.1)
-      return imf,(imf_prev[1]-imf,false)
+        return nothing
+    elseif sum(abs, imf_prev[1]) > 0.0 && !ismonotonic(imf_prev[1])
+        imf = sift(imf_prev[1], iter.xvec, 0.1)
+        return imf, (imf_prev[1]-imf, false)
     else
-      return imf_prev[1],(imf_prev[1],true)
+        return imf_prev[1], (imf_prev[1], true)
     end
 end
 
 Base.IteratorSize(::Type{EMDIterable{U,V}}) where {U,V} = Base.SizeUnknown()
 
 
 """
-emd(measurements, xvec)
+    emd(measurements, xvec, numimfs=6)
 
-Return the Intrinsic Mode Functions and
-the residual of the Empirical Mode Decomposition of the measurements given on time steps given in xvec.
+Return the Intrinsic Mode Functions (IMF) and the residual of the Empirical Mode
+Decomposition (EMD) of the `measurements` given on time steps given in `xvec`.
 """
-function emd(measurements,xvec, num_imfs=6)
-    collect(take(EMDIterable(measurements,xvec),num_imfs))
+function emd(measurements::Vector{T}, xvec, num_imfs=6) where T<:AbstractFloat
```

Comment:
`xvec` can still be a range/iterator in this case, but `measurements` does not make sense to be an iterator.

Reply:
It would make sense to be able to put Iterators into this function. The first example that comes to mind is [skipmissing](https://docs.julialang.org/en/v1/base/base/#Base.skipmissing). This is an iterator and we wouldn't need to collect it to put it into the emd function. For a longer explanation why we don't need to constrain type parameters see this blog post especially the section about over-constraining argument types. 
https://www.oxinabox.net/2020/04/19/Julia-Antipatterns.html

Other number types that are not subtypes of AbstractFloat that might be interesting to use are for example Measurements.jl or Unitful.jl to track either measurement uncertainties or units through the computations. Non AbstractFloat numbers are also used for automatic differentiation.

Reply:
This is a nice post: I learned more about how dispatch works!

My original thought was just about avoiding any in the return element type. If that won't cause any performance issue, in fact I don't mind staying with the current approach.